### PR TITLE
[mqtt.homeassistant] Implement Vacuum discovery for Homeassistant MQTT

### LINKED ISSUE #8988: [mqtt] Support for Home Assistant Component: Vacuum
## Problem
OpenHAB supports already a couple of the "offical" Home Assistant mqtt schemes that many devices follow. But not all of them: The Xioami/Roborock vacuum robots are used by quite a lot of people and most just use the miio binding (which is great btw). But in order to get rid of the cloud used by the robot (for privacy, independence ande feature extension reasons) one can use  [Valetudo.](https://github.com/Hypfer/Valetudo). Valetudo offers mqtt for the robot with the implementation of the Home Assistant  vaccum-scheme. Unfortunately, this scheme is not implemented in openHAB yet.

## Solution
Integration of the [Home Assistant vacuum-scheme](https://www.home-assistant.io/integrations/vacuum.mqtt/) in the openHAB mqtt binding. 

I am far away of knowing enough java to implement this myself, but maybe there is somebody willing to help with this. 

## Your Environment

* Version used: openHAB 2.5.10 Release Build
* Environment name and version (e.g. Chrome 76, Java 8, Node.js 12.9, ...): docker:latest
* Operating System and version (desktop or mobile, Windows 10, Raspbian Buster, ...): Ubuntu 20.04

### PR DESCRIPTION
**Pull request by t2000**
Closes #8988

Automatically discovers vacuum robots running on Valetudo (RE) with enabled MQTT

Signed-off-by: Stefan Triller <github@stefantriller.de>

### CODE REVIEW

**Comment by t2000** [comment]
@fwolter I have rebased my changes and resolved the conflict.

File: bundles/org.openhab.binding.mqtt.homeassistant/src/main/java/org/openhab/binding/mqtt/homeassistant/internal/component/Vacuum.java [hunk 1]
```diff
@@ -0,0 +1,265 @@
...
+import org.openhab.binding.mqtt.homeassistant.internal.config.dto.AbstractChannelConfiguration;
+
+/**
+ * A MQTT vacuum, following the https://www.home-assistant.io/components/vacuum.mqtt/ specification.
+ *
+ * @author Stefan Triller - Initial contribution
+ */
+@NonNullByDefault
+public class Vacuum extends AbstractComponent<Vacuum.ChannelConfiguration> {
+    public static final String vacuumStateChannelID = "state";
```
Comment (fwolter) [inline bundles/org.openhab.binding.mqtt.homeassistant/src/main/java/org/openhab/binding/mqtt/homeassistant/internal/component/Vacuum.java:34]: Constants should be all upper case.

File: bundles/org.openhab.binding.mqtt.homeassistant/src/main/java/org/openhab/binding/mqtt/homeassistant/internal/component/Vacuum.java [hunk 2]
```diff
@@ -0,0 +1,265 @@
...
+    /**
+     * Configuration class for MQTT component
+     */
+    static class ChannelConfiguration extends AbstractChannelConfiguration {
+        ChannelConfiguration() {
+            super("MQTT Vacuum");
+        }
+
+        protected @Nullable String command_topic;
+        protected String state_topic = "";
```
Comment (fwolter) [inline bundles/org.openhab.binding.mqtt.homeassistant/src/main/java/org/openhab/binding/mqtt/homeassistant/internal/component/Vacuum.java:73]: Normal members should be camelCase.

**Review by fwolter — CHANGES_REQUESTED** [review]

**Comment by t2000** [comment]
Thanks for the review, I have fixed the format error as well and rebased against the latest `main` branch too.

**Review by fwolter — APPROVED** [review]

File: bundles/org.openhab.binding.mqtt.homeassistant/src/main/java/org/openhab/binding/mqtt/homeassistant/internal/component/Vacuum.java [hunk 3]
```diff
@@ -0,0 +1,264 @@
...
+    /**
+     * Configuration class for MQTT component
+     */
+    static class ChannelConfiguration extends AbstractChannelConfiguration {
+        ChannelConfiguration() {
+            super("MQTT Vacuum");
+        }
+
+        protected @Nullable String commandTopic;
+        protected String stateTopic = "";
```
Comment (jochen314) [inline bundles/org.openhab.binding.mqtt.homeassistant/src/main/java/org/openhab/binding/mqtt/homeassistant/internal/component/Vacuum.java:73]: Does this work?
Don't we need a @SerializedName("state_topic") vor the deserialization to work?

**Comment by antroids** [comment]
> Closes #8988
> 
> Automatically discovers vacuum robots running on Valetudo (RE) with enabled MQTT
> 
> Signed-off-by: Stefan Triller [github@stefantriller.de](mailto:github@stefantriller.de)

Hi, 
thank you for your contribution!

Could you please show where you found specification of status attributes, like time or usage counters? 
I'm not 100% sure, but looks like these attributes were removed in latest Valetudo versions and they still exists only for Homie. Unfortunately, I'm far away from home and can't check this. 
There is also lack of full support of Home Assistant mqtt.vacuum specification in your implementation.

I have a branch with mqtt.vacuum specification fully implemented: https://github.com/antroids/openhab-addons/tree/mqtt_vacuum
We could merge status attributes into the branch, if there is specification on them somewhere.

**Comment by t2000** [comment]
>Could you please show where you found specification of status attributes, like time or usage counters?

I have used https://www.home-assistant.io/integrations/vacuum.mqtt/ and he output of my valetudo running vacuum cleaners :)

> I'm not 100% sure, but looks like these attributes were removed in latest Valetudo versions and they still exists only for Homie. Unfortunately, I'm far away from home and can't check this.

You are right! Just when I had finished my implementation and when I was searching for possibilities to overcome the map rotation issues in the robots, I also saw that valetudo has changed the mqtt schema to homie. But that shouldn't be an issue, since we could also implement this too.

> There is also lack of full support of Home Assistant mqtt.vacuum specification in your implementation.
my implementation was never meant to cover 100% of any spec. I built it to support the robots out of the box with all the features I need. Feel free to add more, it is OSS.

> I have a branch with mqtt.vacuum specification fully implemented: https://github.com/antroids/openhab-addons/tree/mqtt_vacuum
> We could merge status attributes into the branch, if there is specification on them somewhere.

I didn't know that? How can one know this if you do not leave a comment on the issue that I have closed? and why did you not create a PR yourself earlier or have commented on mine?

Anyway, feel free to add your stuff from your branch, which is missing in my implementation. I mean, in the end this is how OSS works, right? :)

**Comment by antroids** [comment]
Maybe you have a mqtt messages from new Valetudo version? 
I would like to figure out how to handle these schema extensions, because they are specific for particular firmware version.
Maybe enable them based on manufacturer and topic prefix.

**Pull request merged**